OutOfSchool/OutOfSchool.WebApi.Tests/Services/ApplicationSensitiveServiceTests.cs

```diff
@@ -0,0 +1,359 @@
+        currentUserServiceMock.Setup(c => c.IsRegionAdmin()).Returns(true);
+        regionAdminServiceMock
+            .Setup(m => m.GetByUserId(It.IsAny<string>()))
+            .Returns(Task.FromResult<RegionAdminDto>(new RegionAdminDto()
+            {
+                InstitutionId = institutionId,
+                CATOTTGId = catottgId,
+            }));
+
+        currentUserServiceMock.Setup(c => c.IsRegionAdmin()).Returns(true);
+        regionAdminServiceMock
+            .Setup(m => m.GetByUserId(It.IsAny<string>()))
+            .Returns(Task.FromResult<RegionAdminDto>(new RegionAdminDto()
+            {
+                InstitutionId = institutionId,
+            }));
+
+        codeficatorServiceMock
+            .Setup(x => x.GetAllChildrenIdsByParentIdAsync(It.IsAny<long>()))
+            .Returns(Task.FromResult((IEnumerable<long>)new List<long> { catottgId }));
+
+        // Act
+        var result = await service.GetAll(new ApplicationFilter());
+
+        // Assert
+        Assert.That(result.Entities.Count, Is.EqualTo(1));
+        Assert.That(result.Entities.FirstOrDefault().Workshop.InstitutionId, Is.EqualTo(institutionId));
+    }
+
+    [Test]
+    public async Task GetApplications_WhenAreaAdminCalled_ShouldReturnApplications()
+    {
+        // Arrange
+        var institutionId = new Guid("b929a4cd-ee3d-4bad-b2f0-d40aedf656c4");
+        long catottgId = 31737;
+        var applications = WithApplicationsList();
+        SetupGetAllByInstitutionId(applications);
+
+        currentUserServiceMock.Setup(c => c.IsAdmin()).Returns(true);
+        currentUserServiceMock.Setup(c => c.IsMinistryAdmin()).Returns(false);
+        currentUserServiceMock.Setup(c => c.IsRegionAdmin()).Returns(false);
+        currentUserServiceMock.Setup(c => c.IsAreaAdmin()).Returns(true);
+        areaAdminServiceMock
+            .Setup(m => m.GetByUserId(It.IsAny<string>()))
+            .Returns(Task.FromResult<AreaAdminDto>(new AreaAdminDto()
+            {
+                InstitutionId = institutionId,
+                CATOTTGId = catottgId,
+            }));
+
+        currentUserServiceMock.Setup(c => c.IsAreaAdmin()).Returns(true);
+        areaAdminServiceMock
+            .Setup(m => m.GetByUserId(It.IsAny<string>()))
+            .Returns(Task.FromResult<AreaAdminDto>(new AreaAdminDto()
+            {
+                InstitutionId = institutionId,
+            }));
+
+        codeficatorServiceMock
+            .Setup(x => x.GetAllChildrenIdsByParentIdAsync(It.IsAny<long>()))
+            .Returns(Task.FromResult((IEnumerable<long>)new List<long> { catottgId }));
+
+        // Act
+        var result = await service.GetAll(new ApplicationFilter());
+
+        // Assert
+        Assert.That(result.Entities.Count, Is.EqualTo(1));
+        Assert.That(result.Entities.FirstOrDefault().Workshop.InstitutionId, Is.EqualTo(institutionId));
+    }
+
+    private List<Application> WithApplicationsList()
```

Comment:
(Tests readability, fragility) This approach - to provide test data having specific entities for every test is suboptimal:
1. It's unclear which constants could be changed (b/c we don't check them in tests) and which have dependencies in test code.
2. (Similar to previous) It makes tests obscure (i.e. there is line `Assert.That(result.Entities.Count, Is.EqualTo(1));` where 1 is a magic number and noone knows why it is there)
3. It makes hard to add new tests, because while adding more data entities or changing existing you could break other tests.
4. It doesn't use all power of fake test data generators.

Suggested change
You could still have `WithApplicationsList()` method that generates test-neutral data (though it could be easily replaced by `faker.Generate()`). 
In arrange phase - you take this data for your test, change N (and N could even be random, no magical 1s) entities to match test criteria, perform act phase as usual and assert that expected N is obtained in results.

Reply:
Resolved but nothing done.

Reply:
> Resolved but nothing done.

@VadymLevkovskyi Can you please a little bit deeper about this comment. I don`t full understand how it should be look like. I tried to introduce a field which was like a constant parament ```N``` and inside this test I replaced hardcode number with these value of ```N```. But as I understand now it should be done in another way. And I don`t understand what should be done with this paramert. Is it should be used like an attribute. Because when I mock smth I don`t quite understand how to return exactly N element and how it should be written there.
Thank you in advance!)

Reply:
Little tiny bit deeper ;) 

Few comments about code below:
1. Added test for not admin case 
2. All data are taken from generators
3. Wrong data are explicitely constructed in related test, no magic numbers used, no generic "wrong" data created for all tests - thus tests are more independent.
4. As many as possible replacements of `It.IsAny` is done - thus we see what ids are expected in what methods
5. `Assert.AreEqual(result.Entities.Count, application.Count);` is dropped everywhere, because it doesn't check the SUT implementation, instead it tests part of the same test code (specifically `mapper.Setup`), so there is no reason to reverify it. Rephrasing, that assert is schematically same to 
```
// somewhere in test code
var isTrue = true;
...
Assert.That(isTrue. Is.True());
```
6. Pay attention to the real assertion - powered by .It.Is<Expression<...>>(...) and related .VerifyAll()

Sorry for lot of text. Hope it will be helpful.

```
using AutoMapper;
using Microsoft.Extensions.Localization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Moq;
using NUnit.Framework;
using OutOfSchool.EmailSender;
using OutOfSchool.RazorTemplatesData.Services;
using OutOfSchool.Services.Enums;
using OutOfSchool.Services.Models;
using OutOfSchool.Services.Repository;
using OutOfSchool.Tests.Common;
using OutOfSchool.Tests.Common.TestDataGenerators;
using OutOfSchool.WebApi.Config;
using OutOfSchool.WebApi.Models;
using OutOfSchool.WebApi.Models.Application;
using OutOfSchool.WebApi.Services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
namespace OutOfSchool.WebApi.Tests.Services;

[TestFixture]
public class ApplicationSensitiveServiceTests
{
    private ISensitiveApplicationService service;
    private string currentUserId;
    private Mock<IApplicationRepository> applicationRepositoryMock;
    private Mock<IMapper> mapper;
    private Mock<ICurrentUserService> currentUserServiceMock;
    private Mock<IMinistryAdminService> ministryAdminServiceMock;
    private Mock<IRegionAdminService> regionAdminServiceMock;
    private Mock<IAreaAdminService> areaAdminServiceMock;
    private Mock<ICodeficatorService> codeficatorServiceMock;

    [SetUp]
    public void SetUp()
    {
        applicationRepositoryMock = new Mock<IApplicationRepository>();
        mapper = new Mock<IMapper>();
        currentUserServiceMock = new Mock<ICurrentUserService>();
        ministryAdminServiceMock = new Mock<IMinistryAdminService>();
        regionAdminServiceMock = new Mock<IRegionAdminService>();
        areaAdminServiceMock = new Mock<IAreaAdminService>();
        codeficatorServiceMock = new Mock<ICodeficatorService>();

        service = new ApplicationService(
            applicationRepositoryMock.Object,
            new Mock<ILogger<ApplicationService>>().Object,
            new Mock<IWorkshopRepository>().Object,
            mapper.Object,
            new Mock<IOptions<ApplicationsConstraintsConfig>>().Object,
            new Mock<INotificationService>().Object,
            new Mock<IProviderAdminService>().Object,
            new Mock<IChangesLogService>().Object,
            new Mock<IWorkshopServicesCombiner>().Object,
            currentUserServiceMock.Object,
            ministryAdminServiceMock.Object,
            regionAdminServiceMock.Object,
            areaAdminServiceMock.Object,
            codeficatorServiceMock.Object,
            new Mock<IRazorViewToStringRenderer>().Object,
            new Mock<IEmailSender>().Object,
            new Mock<IStringLocalizer<SharedResource>>().Object,
            new Mock<IOptions<HostsConfig>>().Object);

        currentUserId = Guid.NewGuid().ToString();
        currentUserServiceMock.SetupGet(c => c.UserId).Returns(currentUserId);
    }

    [Test]
    public void GetAll_WhenNotAdmin_Throws()
    {
        // Arrange
        SetupAdminRights(isAdmin: false);

        // Act, Assert
        Assert.ThrowsAsync<UnauthorizedAccessException>(async () => await service.GetAll(new ApplicationFilter()));
    }

    [Test]
    public async Task GetAll_WhenMinistryAdmin_ReturnsApplications()
    {
        // Arrange
        var applications = GenerateApplicationsForSameWorkshop();

        SetupAdminRights(isAdmin: true, isMinistry: true);

        var appFilter = new ApplicationFilter();
        SetupRepositoryWithVerification(appFilter, applications);

        ministryAdminServiceMock.Setup(m => m.GetByIdAsync(currentUserId))
            .Returns(Task.FromResult(new MinistryAdminDto() { InstitutionId = applications[0].Workshop.InstitutionHierarchy.InstitutionId }));

        // Act
        var result = await service.GetAll(appFilter);

        // Assert
        mapper.VerifyAll();
        currentUserServiceMock.VerifyAll();
        applicationRepositoryMock.VerifyAll();
    }

    [Test]
    public async Task GetAll_WhenMinistryAdmin_ReturnsRelatedApplications()
    {
        // Arrange
        var applications = GenerateApplicationsForSameWorkshop();
        var institutionId = applications[0].Workshop.InstitutionHierarchy.InstitutionId;

        var mnistryAdmin = AdminGenerator.GenerateMinistryAdminDto();
        mnistryAdmin.InstitutionId = institutionId;

        var otherInstitutionHierarchy = InstitutionHierarchyGenerator.Generate();
        var otherWorkshop = WorkshopGenerator.Generate().WithInstitutionHierarchy(otherInstitutionHierarchy);

        var wrongApplications = applications.Take(1).ToList().WithWorkshop(otherWorkshop);
        var correctApplications = applications.Skip(wrongApplications.Count).ToList();

        SetupAdminRights(isAdmin: true, isMinistry: true);

        ministryAdminServiceMock.Setup(m => m.GetByIdAsync(currentUserId)).Returns(Task.FromResult(mnistryAdmin));

        var appFilter = new ApplicationFilter();
        SetupRepositoryWithVerification(appFilter, correctApplications, wrongApplications);

        // Act
        var result = await service.GetAll(appFilter);

        // Assert
        mapper.VerifyAll();
        currentUserServiceMock.VerifyAll();
        ministryAdminServiceMock.VerifyAll();
        applicationRepositoryMock.VerifyAll();
    }

    [Test]
    public async Task GetAll_WhenRegionAdmin_ReturnsRelatedApplications()
    {
        // Arrange
        var applications = GenerateApplicationsForSameWorkshop();
        var workshop = applications[0].Workshop;
        var institutionId = workshop.InstitutionHierarchy.InstitutionId;
        var catottgId = workshop.Provider.LegalAddress.CATOTTGId;

        var regionAdmin = AdminGenerator.GenerateRegionAdminDto();
        regionAdmin.InstitutionId = institutionId;

        var childrenCatottgIds = new List<long> { catottgId } as IEnumerable<long>;

        var otherInstitutionHierarchyWorkshop = WorkshopGenerator.Generate().WithInstitutionHierarchy(InstitutionHierarchyGenerator.Generate()).WithProvider(workshop.Provider);
        var otherProviderWorkshop = WorkshopGenerator.Generate().WithInstitutionHierarchy(workshop.InstitutionHierarchy).WithProvider(ProvidersGenerator.Generate());

        var wrongApplications = new List<Application> 
        {
            applications[0].WithWorkshop(otherInstitutionHierarchyWorkshop),
            applications[1].WithWorkshop(otherProviderWorkshop),
        };
        var correctApplications = applications.Skip(wrongApplications.Count).ToList();

        SetupAdminRights(isAdmin: true, isRegion: true);

        var appFilter = new ApplicationFilter();
        SetupRepositoryWithVerification(appFilter, correctApplications, wrongApplications);

        regionAdminServiceMock.Setup(m => m.GetByUserId(currentUserId)).Returns(Task.FromResult(regionAdmin));

        codeficatorServiceMock
            .Setup(x => x.GetAllChildrenIdsByParentIdAsync(regionAdmin.CATOTTGId))
            .Returns(Task.FromResult(childrenCatottgIds));

        // Act
        var result = await service.GetAll(appFilter);

        // Assert
        mapper.VerifyAll();
        currentUserServiceMock.VerifyAll();
        regionAdminServiceMock.VerifyAll();
        codeficatorServiceMock.VerifyAll();
        applicationRepositoryMock.VerifyAll();
    }

    [Test]
    public async Task GetAll_WhenAreaAdmin_ReturnsRelatedApplications()
    {
        // Arrange
        var applications = GenerateApplicationsForSameWorkshop();
        var workshop = applications[0].Workshop;
        var institutionId = workshop.InstitutionHierarchy.InstitutionId;
        var catottgId = workshop.Provider.LegalAddress.CATOTTGId;

        var areaAdmin = AdminGenerator.GenerateAreaAdminDto();
        areaAdmin.InstitutionId = institutionId;

        var childrenCatottgIds = new List<long> { catottgId } as IEnumerable<long>;

        var otherInstitutionHierarchyWorkshop = WorkshopGenerator.Generate().WithInstitutionHierarchy(InstitutionHierarchyGenerator.Generate()).WithProvider(workshop.Provider);
        var otherProviderWorkshop = WorkshopGenerator.Generate().WithInstitutionHierarchy(workshop.InstitutionHierarchy).WithProvider(ProvidersGenerator.Generate());

        var wrongApplications = new List<Application> 
        {
            applications[0].WithWorkshop(otherInstitutionHierarchyWorkshop),
            applications[1].WithWorkshop(otherProviderWorkshop),
        };
        var correctApplications = applications.Skip(wrongApplications.Count).ToList();

        SetupAdminRights(isAdmin: true, isArea: true);

        var appFilter = new ApplicationFilter();
        SetupRepositoryWithVerification(appFilter, correctApplications, wrongApplications);

        areaAdminServiceMock.Setup(m => m.GetByUserId(currentUserId)).Returns(Task.FromResult(areaAdmin));

        codeficatorServiceMock
            .Setup(x => x.GetAllChildrenIdsByParentIdAsync(areaAdmin.CATOTTGId))
            .Returns(Task.FromResult(childrenCatottgIds));

        // Act
        var result = await service.GetAll(appFilter);

        // Assert
        mapper.VerifyAll();
        currentUserServiceMock.VerifyAll();
        areaAdminServiceMock.VerifyAll();
        codeficatorServiceMock.VerifyAll();
        applicationRepositoryMock.VerifyAll();
    }

    private void SetupAdminRights(bool isAdmin, bool isMinistry = false, bool isRegion = false, bool isArea = false)
    {
        currentUserServiceMock.Setup(c => c.IsAdmin()).Returns(isAdmin);
        currentUserServiceMock.Setup(c => c.IsMinistryAdmin()).Returns(isMinistry);
        currentUserServiceMock.Setup(c => c.IsRegionAdmin()).Returns(isRegion);
        currentUserServiceMock.Setup(c => c.IsAreaAdmin()).Returns(isArea);
    }

    private List<Application> GenerateApplicationsForSameWorkshop()
    {
        var institutionHierarchy = InstitutionHierarchyGenerator.Generate();
        var workshop = WorkshopGenerator.Generate().WithInstitutionHierarchy(institutionHierarchy).WithProvider();
        var parent = ParentGenerator.Generate();
        var child = ChildGenerator.Generate().WithParent(parent);

        return ApplicationGenerator.Generate(6).WithWorkshop(workshop).WithParent(parent).WithChild(child);
    }

    private void SetupRepositoryWithVerification(ApplicationFilter appFilter, List<Application> correctApplications, List<Application> wrongApplications = null)
    {
        applicationRepositoryMock.Setup(
            w => w.Get(
                appFilter.From,
                appFilter.Size,
                It.IsAny<string>(),
                It.Is<Expression<Func<Application, bool>>>(
                    // here main magic goes - verification that predicate built in service and passed to applicationRepository should behave correctly on "correct" and "wrong" applications
                    expr => correctApplications.All(expr.Compile())
                        && (wrongApplications == null || !wrongApplications.Any(expr.Compile()))
                ),
                It.IsAny<Dictionary<Expression<Func<Application, object>>, SortDirection>>(),
                It.IsAny<bool>()))
            .Returns(correctApplications.AsTestAsyncEnumerableQuery())
            .Verifiable("Wrong expression was built in service and passed to applicationRepository.Get() whereExpression argument. Probably some conditions aren't tested");

        mapper.Setup(m => m.Map<List<ApplicationDto>>(It.IsAny<List<Application>>()))
            .Returns(correctApplications.Select(app => new ApplicationDto()).ToList());
    }
}
```

Also improved AddressGenerator:
```
    static AddressGenerator()
    {
        faker = new Faker<Address>()
            .RuleFor(x => x.Id, f => f.IndexFaker)
            .RuleFor(x => x.CATOTTG, _ => CATOTTGGenerator.Generate())
            .RuleFor(x => x.CATOTTGId, (f, address) => address.CATOTTG.Id) // <-- this line was changed
            .RuleFor(x => x.Street, f => f.Address.StreetName())
            .RuleFor(x => x.BuildingNumber, f => f.Address.BuildingNumber())
            .RuleFor(x => x.Latitude, f => f.Address.Latitude())
            .RuleFor(x => x.Longitude, f => f.Address.Longitude());

        // Increment initial value of IndexFaker to have first created entity with Id=1
        // and prevent System.InvalidOperationException when it is added to the DbContext
        (faker as IFakerTInternal).FakerHub.IndexFaker++;
    }
```